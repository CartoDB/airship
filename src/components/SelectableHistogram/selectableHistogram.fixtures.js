export default [
  {
    start: 8,
    end: 48,
    value: 6318,
  },
  {
    start: 48,
    end: 88,
    value: 5922,
  },
  {
    start: 88,
    end: 128,
    value: 2363,
  },
  {
    start: 128,
    end: 168,
    value: 801,
  },
  {
    start: 168,
    end: 208,
    value: 441,
  },
  {
    start: 208,
    end: 248,
    value: 118,
  },
  {
    start: 248,
    end: 288,
    value: 92,
  },
  {
    start: 288,
    end: 328,
    value: 77,
  },
  {
    start: 328,
    end: 368,
    value: 48,
  },
  {
    start: 368,
    end: 408,
    value: 34,
  },
  {
    start: 408,
    end: 448,
    value: 10,
  },
  {
    start: 448,
    end: 488,
    value: 20,
  },
  {
    start: 488,
    end: 528,
    value: 16,
  },
  {
    start: 528,
    end: 568,
    value: 6,
  },
  {
    start: 568,
    end: 608,
    value: 10,
  },
  {
    start: 608,
    end: 648,
    value: 2,
  },
];
