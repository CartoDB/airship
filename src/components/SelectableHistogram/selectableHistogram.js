import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { darken } from 'polished';
import { select, event } from 'd3-selection';
import { scaleLinear } from 'd3-scale';
import { max, range } from 'd3-array';
import { brushX } from 'd3-brush';
import { axisLeft } from 'd3-axis';
import 'd3-transition';
import Base from '../Typography/base';
import { theme } from '../../constants';
import { readableNumber } from '../../utils';

const WIDTH = 205;
const HEIGHT = 125;
const DIVISION_WIDTH = 80;
const BARS_SEPARATION = 1;
const MARGIN = {
  TOP: 15,
  RIGHT: 3,
  BOTTOM: 20,
  LEFT: 40,
};

const Tooltip = styled.div`
  position: absolute;
  background: ${props => props.theme.black};
  border-radius: 4px;
  padding: 0.5rem;
  opacity: 0;
  z-index: 100;
`;
Tooltip.displayName = 'Tooltip';
Tooltip.defaultProps = {
  theme,
};

const TooltipValue = Base.extend`
  font-size: 14px;
  line-height: 14px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
  width: 100%;
  cursor: default;
  color: ${props => props.theme.white};
`;
TooltipValue.defaultProps = {
  theme,
};

const Svg = styled.svg.attrs({
  viewBox: '0 0 248 160',
})`
  .tick {
    line {
      opacity: 0.1;
      stroke: ${props => props.theme.type02};
    }

    :first-child {
      line {
        opacity: 0.5;
      }
    }
  }

  .bar {
    fill: ${props => props.color || props.theme.brand03};
    cursor: pointer;
    opacity: 0.7;
    transition: all 0.3s linear;

    :hover {
      fill: ${props => darken(0.16, props.color || props.theme.brand03)};
    }
  }

  text {
    font-size: 10px;
    font-weight: 300;
    fill: ${props => props.theme.type01};
    font-family: Roboto, sans-serif};
    -webkit-font-smoothing: antialiased;
  }
`;
Svg.defaultProps = {
  theme,
};

class SelectableHistogram extends Component {
  static defaultProps = {
    data: [],
  };

  static propTypes = {
    color: PropTypes.string,
    data: PropTypes.arrayOf(PropTypes.shape({
      start: PropTypes.number,
      end: PropTypes.number,
      value: PropTypes.number,
    })),
  };

  state = {
    tooltip: null,
  }

  componentDidMount() {
    this.renderYAxis();
    this.renderXAxis();

    this.barsContainer = this.container
      .append('g')
      .attr('transform', `translate(${MARGIN.LEFT}, ${MARGIN.TOP})`);
    const self = this;
    this.brushContainer = this.container
        .append('g')
        .attr("class", "brush")
        .attr('transform', `translate(${MARGIN.LEFT}, ${MARGIN.TOP})`)
        .call(brushX()
            .extent([[0, 0], [WIDTH, HEIGHT]])
            .on("brush",  function ()  {
                if (event.sourceEvent.type === "brush") return;
                const d0 = event.selection.map(self.xScale.invert);
                console.log('data', self.props.data);
                const d1 = d0.map(value => {
                    const element = self.props.data.find(bar => value > bar.start && value < bar.end);
                    return element ? element.start : value;
                });
                if (self.props.onSelectedData) {
                    self.props.onSelectedData({ min:d1[0], max:d1[1]});
                };

                select(this).call(event.target.move, d1.map(self.xScale));
            }));
    this.renderBars();
  }

  componentDidUpdate() {
    this.updateAxis();
    this.renderBars();
  }

  renderXAxis() {
    const { data } = this.props;
    const { start } = data[0];
    const { end } = data[data.length - 1];
    const divisions = Math.round(WIDTH / DIVISION_WIDTH);
    const step = WIDTH / divisions;
    const stop = WIDTH + step;

    // -- X Axis
    this.xScale = scaleLinear()
      .domain([start, end])
      .range([0, WIDTH]);

    this.xAxis = range(0, stop, step).slice(0, divisions + 1);

    this.xAxisSelection = this.container
      .append('g')
      .attr('transform', `translate(${MARGIN.LEFT}, ${MARGIN.TOP})`)
      .selectAll('text')
      .data(this.xAxis)
      .enter()
      .append('text')
      .attr('x', d => d)
      .attr('y', () => HEIGHT + 15)
      .attr('text-anchor', (d, index) => {
        if (index === 0) return 'start';
        if (index === (this.xAxis.length - 1)) return 'end';

        return 'middle';
      })
      .text(d => readableNumber(this.xScale.invert(d)));
  }

  renderYAxis() {
    const { data } = this.props;

    // -- Y Axis
    this.yScale = scaleLinear()
      .range([HEIGHT, 0])
      .domain([0, max(data, d => d.value)])
      .nice();

    this.yAxis = axisLeft(this.yScale)
      .tickSize(-WIDTH, 0, 0)
      .ticks(5);

    this.yAxisSelection = this.container
      .append('g')
      .attr('transform', `translate(${MARGIN.LEFT}, ${MARGIN.TOP})`)
      .call(this.yAxis);

    select('.domain').remove(); // Remove axis border
  }

  renderBars() {
    const { data } = this.props;
    const barWidth = WIDTH / data.length;

    // -- Draw bars
    this.bars = this.barsContainer
      .selectAll('rect')
      .data(data);

    // -- Exit
    this.bars.exit().remove();

    // -- Enter
    this.bars
      .enter()
      .append('rect')
      .on('mouseout', () => this.setState({ tooltip: null }))
      .on('mouseenter', d => {
        this.setState(
          { tooltip: { d } },
          () => this.showTooltip(event.layerX, event.layerY)
        );
      })
      .on('mousemove', () => {
        select(this.tooltipNode).style('opacity', 0);
        this.showTooltip(event.layerX, event.layerY);
      })
      .merge(this.bars)
      .attr('class', 'bar')
      .attr('y', HEIGHT)
      .attr('x', (d, index) => index * barWidth)
      .attr('width', () => Math.max(0, barWidth - BARS_SEPARATION))
      .attr('height', 0)
      .transition()
      .delay(200)
      .attr('y', d => this.yScale(d.value))
      .attr('height', d => HEIGHT - this.yScale(d.value));

    // -- Update
    this.bars
      .attr('y', d => this.yScale(d.value))
      .attr('height', d => HEIGHT - this.yScale(d.value));
  }

  updateAxis() {
    const { data } = this.props;
    const { start } = data[0];
    const { end } = data[data.length - 1];

    // -- Update scales
    this.yScale
      .domain([0, max(data, d => d.value)])
      .nice();

    this.xScale
      .domain([start, end]);

    // -- Update axis
    this.xAxisSelection
      .text(d => readableNumber(this.xScale.invert(d)));

    this.yAxisSelection
      .call(this.yAxis);

    select('.domain').remove(); // Remove axis border
  }

  getTooltipPosition(mouseX, mouseY) {
    const OFFSET = 25;
    let x = mouseX;
    let y = mouseY;

    const viewportBoundaries = {
      right: window.innerWidth + window.pageXOffset,
      bottom: window.innerHeight + window.pageYOffset,
    };

    const tooltipContainerBoundingRect = this.tooltipNode.getBoundingClientRect();

    const tooltipBoundaries = {
      right: mouseX + tooltipContainerBoundingRect.width,
      bottom: mouseY + tooltipContainerBoundingRect.height,
    };

    if (viewportBoundaries.right < tooltipBoundaries.right) {
      x = mouseX - tooltipContainerBoundingRect.width;
    }

    if (viewportBoundaries.bottom < tooltipBoundaries.bottom) {
      y = mouseY - tooltipContainerBoundingRect.height - OFFSET;
    }

    return [x, y];
  }

  showTooltip(mouseX, mouseY) {
    const [x, y] = this.getTooltipPosition(mouseX, mouseY);

    select(this.tooltipNode)
      .style('opacity', 1)
      .style('left', `${x + 10}px`)
      .style('top', `${y + 20}px`);
  }

  renderTooltip() {
    const { d } = this.state.tooltip;

    return (
      <Tooltip innerRef={node => { this.tooltipNode = node; }}>
        <TooltipValue>{readableNumber(d.value)}</TooltipValue>
      </Tooltip>
    );
  }

  render() {
    return (
      <React.Fragment>
        <Svg innerRef={node => { this.container = select(node); }} {...this.props} />
        {this.state.tooltip && this.renderTooltip()}
      </React.Fragment>
    );
  }
}

export default SelectableHistogram;
