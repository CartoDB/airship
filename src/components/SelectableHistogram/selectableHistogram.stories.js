import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import SelectableHistogram from './selectableHistogram';
import { action } from '@storybook/addon-actions';
import Widget from '../Widget/widget';
import mockData from './selectableHistogram.fixtures';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  ui01: '#333',
  type01: 'white',
  type02: '#F5F5F5',
};
class HistogramUpdated extends React.Component {
  state = {
    data: mockData,
  }

  changeState = () => {
    this.setState(prevState => ({
      data: prevState.data.length !== 3 ? mockData.slice(2, 5) : mockData,
    }));
  }

  render() {
    return (
      <div>
        <SelectableHistogram data={this.state.data} />
        <button onClick={this.changeState}>Click me</button>
      </div>
    );
  }
}

storiesOf('SelectableHistogram', module)
  .add('selectable', () => (
    <SelectableHistogram data={mockData} selected={{min:8, max:88}} onSelectedData={action('selected')} />
  ))
  .add('Updating data', () => (
    <Widget>
      <Widget.Title>Suffer score</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <HistogramUpdated />
    </Widget>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <Widget>
        <Widget.Title>Suffer score</Widget.Title>
        <Widget.Description>Just a widget</Widget.Description>

        <SelectableHistogram data={mockData} />
      </Widget>
    </ThemeProvider>
  ));
