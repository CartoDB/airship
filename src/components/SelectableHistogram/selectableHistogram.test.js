import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import SelectableHistogram from './selectableHistogram';
import mockData from './selectableHistogram.fixtures';

describe('<SelectableHistogram />', () => {
  describe('render', () => {
    describe('when data passed in', () => {
      it('renders the correct amount of bars', () => {
        const component = renderer.create(<SelectableHistogram data={mockData} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('renders the tooltip', () => {
        const component = mount(<SelectableHistogram data={mockData} />);

        component.setState({ tooltip: { d: { name: 'Leet', value: 1337 } } });
        component.update();

        expect(component.find('Tooltip').length).toEqual(1);
      });
    });
  });

  describe('update', () => {
    describe('when data changes', () => {
      it('should call the update method or the chart', () => {
        const component = renderer.create(<SelectableHistogram data={mockData} />);

        expect(component.toJSON()).toMatchSnapshot();

        // Changing properties should trigger a componentDidUpdate
        component.update(<SelectableHistogram data={[{ start: 0, end: 10, value: 500 }]} />);

        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });
});
