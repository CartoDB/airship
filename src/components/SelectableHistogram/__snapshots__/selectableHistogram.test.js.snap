// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Histogram /> update when data changes should call the update method or the chart 1`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
}

.c0 .tick line {
  opacity: 0.1;
  stroke: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #47DB99;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #22ae70;
}

.c0 text {
  font-size: 10px;
  font-weight: 300;
  fill: #2C2C2C;
  font-family: Roboto,sans-serif;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "end": 48,
        "start": 8,
        "value": 6318,
      },
      Object {
        "end": 88,
        "start": 48,
        "value": 5922,
      },
      Object {
        "end": 128,
        "start": 88,
        "value": 2363,
      },
      Object {
        "end": 168,
        "start": 128,
        "value": 801,
      },
      Object {
        "end": 208,
        "start": 168,
        "value": 441,
      },
      Object {
        "end": 248,
        "start": 208,
        "value": 118,
      },
      Object {
        "end": 288,
        "start": 248,
        "value": 92,
      },
      Object {
        "end": 328,
        "start": 288,
        "value": 77,
      },
      Object {
        "end": 368,
        "start": 328,
        "value": 48,
      },
      Object {
        "end": 408,
        "start": 368,
        "value": 34,
      },
      Object {
        "end": 448,
        "start": 408,
        "value": 10,
      },
      Object {
        "end": 488,
        "start": 448,
        "value": 20,
      },
      Object {
        "end": 528,
        "start": 488,
        "value": 16,
      },
      Object {
        "end": 568,
        "start": 528,
        "value": 6,
      },
      Object {
        "end": 608,
        "start": 568,
        "value": 10,
      },
      Object {
        "end": 648,
        "start": 608,
        "value": 2,
      },
    ]
  }
  viewBox="0 0 248 160"
/>
`;

exports[`<SelectableHistogram /> render when data passed in renders the correct amount of bars 1`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
}

.c0 .tick line {
  opacity: 0.1;
  stroke: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #47DB99;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #22ae70;
}

.c0 text {
  font-size: 10px;
  font-weight: 300;
  fill: #2C2C2C;
  font-family: Roboto,sans-serif;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "end": 48,
        "start": 8,
        "value": 6318,
      },
      Object {
        "end": 88,
        "start": 48,
        "value": 5922,
      },
      Object {
        "end": 128,
        "start": 88,
        "value": 2363,
      },
      Object {
        "end": 168,
        "start": 128,
        "value": 801,
      },
      Object {
        "end": 208,
        "start": 168,
        "value": 441,
      },
      Object {
        "end": 248,
        "start": 208,
        "value": 118,
      },
      Object {
        "end": 288,
        "start": 248,
        "value": 92,
      },
      Object {
        "end": 328,
        "start": 288,
        "value": 77,
      },
      Object {
        "end": 368,
        "start": 328,
        "value": 48,
      },
      Object {
        "end": 408,
        "start": 368,
        "value": 34,
      },
      Object {
        "end": 448,
        "start": 408,
        "value": 10,
      },
      Object {
        "end": 488,
        "start": 448,
        "value": 20,
      },
      Object {
        "end": 528,
        "start": 488,
        "value": 16,
      },
      Object {
        "end": 568,
        "start": 528,
        "value": 6,
      },
      Object {
        "end": 608,
        "start": 568,
        "value": 10,
      },
      Object {
        "end": 648,
        "start": 608,
        "value": 2,
      },
    ]
  }
  viewBox="0 0 248 160"
/>
`;

exports[`<SelectableHistogram /> update when data changes should call the update method or the chart 1`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
}

.c0 .tick line {
  opacity: 0.1;
  stroke: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #47DB99;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #22ae70;
}

.c0 text {
  font-size: 10px;
  font-weight: 300;
  fill: #2C2C2C;
  font-family: Roboto,sans-serif;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "end": 48,
        "start": 8,
        "value": 6318,
      },
      Object {
        "end": 88,
        "start": 48,
        "value": 5922,
      },
      Object {
        "end": 128,
        "start": 88,
        "value": 2363,
      },
      Object {
        "end": 168,
        "start": 128,
        "value": 801,
      },
      Object {
        "end": 208,
        "start": 168,
        "value": 441,
      },
      Object {
        "end": 248,
        "start": 208,
        "value": 118,
      },
      Object {
        "end": 288,
        "start": 248,
        "value": 92,
      },
      Object {
        "end": 328,
        "start": 288,
        "value": 77,
      },
      Object {
        "end": 368,
        "start": 328,
        "value": 48,
      },
      Object {
        "end": 408,
        "start": 368,
        "value": 34,
      },
      Object {
        "end": 448,
        "start": 408,
        "value": 10,
      },
      Object {
        "end": 488,
        "start": 448,
        "value": 20,
      },
      Object {
        "end": 528,
        "start": 488,
        "value": 16,
      },
      Object {
        "end": 568,
        "start": 528,
        "value": 6,
      },
      Object {
        "end": 608,
        "start": 568,
        "value": 10,
      },
      Object {
        "end": 648,
        "start": 608,
        "value": 2,
      },
    ]
  }
  viewBox="0 0 248 160"
/>
`;

exports[`<SelectableHistogram /> update when data changes should call the update method or the chart 2`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
}

.c0 .tick line {
  opacity: 0.1;
  stroke: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #47DB99;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #22ae70;
}

.c0 text {
  font-size: 10px;
  font-weight: 300;
  fill: #2C2C2C;
  font-family: Roboto,sans-serif;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "end": 10,
        "start": 0,
        "value": 500,
      },
    ]
  }
  viewBox="0 0 248 160"
/>
`;
